corehq/motech/repeaters/migrations/0002_repeaters_db.py

```diff
@@ -0,0 +1,170 @@
+# Generated by Django 3.2.16 on 2023-04-12 20:05
+#
+# This migration assumes that repeaters_repeatrecord and
+# repeaters_repeatrecordattempt tables are not in use.
+
+from django.conf import settings
+from django.db import connections, migrations
+from django.db.utils import DEFAULT_DB_ALIAS
+
+# Use a separate user for the fdw connection because postgres_fdw
+# changes session settings like search_path that interfere with other
+# clients because of pg_bouncer connection pooling.
+# https://www.postgresql.org/docs/9.4/postgres-fdw.html#AEN130855
+# PgBouncer maintains a pool of connections for each unique user,
+# database pair.
+# https://pgdash.io/blog/pgbouncer-connection-pool.html
+
+MOVE_TO_SEPARATE_DB = """
+    CREATE SERVER repeaters_fdw FOREIGN DATA WRAPPER postgres_fdw
+        OPTIONS (host '{HOST}', port '{PORT}', dbname '{NAME}');
+    CREATE USER MAPPING FOR {USER} SERVER repeaters_fdw
+        OPTIONS (user 'repeaters_fdw_user', password '{PASSWORD}');
+    GRANT USAGE ON FOREIGN SERVER repeaters_fdw TO {USER};
+
+    DROP TABLE repeaters_repeatrecordattempt;
+    DROP TABLE repeaters_repeatrecord;
+    LOCK TABLE repeaters_repeater IN SHARE ROW EXCLUSIVE mode;
+    ALTER TABLE repeaters_repeater RENAME TO repeaters_repeater_old;
+    IMPORT FOREIGN SCHEMA public
+        LIMIT TO ("repeaters_repeater", "repeaters_repeatrecord", "repeaters_repeater_id_seq_view")
+        FROM SERVER repeaters_fdw INTO public;
+    CREATE FUNCTION repeaters_repeater_id_seq_nextval() RETURNS bigint AS
+        'SELECT value FROM repeaters_repeater_id_seq_view;' LANGUAGE SQL;
+    ALTER TABLE repeaters_repeater ALTER COLUMN id SET DEFAULT repeaters_repeater_id_seq_nextval();
+    INSERT INTO repeaters_repeater SELECT * FROM repeaters_repeater_old;
+    UPDATE repeaters_repeater_id_seq_view SET value = (SELECT MAX(id) + 1 FROM repeaters_repeater);
+    DROP TABLE repeaters_repeater_old;
+"""
+
+# NOTE this does not copy data in repeaters_repeatrecord and
+# repeaters_repeatrecordattempt from the foreign tables. Those tables are
+# expected to contain no data when the forward migration is performed,
+# and that state will be preserved if the migration is reversed. If they
+# did contain data it would likely be too much to copy in a single
+# database transaction, and locking the table in the foreign database is
+# not possible from here anyway.
+#
+# In case it was not clear, this is not a safe migration to run if there
+# are any active writers on the foreign table(s).
+UNDO_MOVE_TO_SEPARATE_DB = """
```

Comment:
Has this rollback been tested?

Reply:
Yes, I tested it on staging during the rollout of https://github.com/dimagi/commcare-cloud/pull/5989